ultralist/input_parser.go

```diff
@@ -0,0 +1,195 @@
+package ultralist
+
+import (
+	"regexp"
+	"strings"
+	"time"
+)
+
+// InputParser parses text to extract a Filter struct
+type InputParser struct{}
+
+/*
+
+# status of not now
+status:-now
+
+# status of now OR next
+status:now,next
+
+# status not now
+status:-now
+status:next
+
+# due today OR tomorrow
+due:tod,tom
+due:today due:tom
+
+# filter todos before a certain date
+due:<aug15
+
+completed:true
+
+priority:false
+
+project:one,-two
+
+*/
+
+// Parse parses raw input and returns a Filter object
+func (p *InputParser) Parse(input string) (*Filter, error) {
+	filter := &Filter{
+		HasStatus:        false,
+		HasCompleted:     false,
+		HasCompletedAt:   false,
+		HasIsPriority:    false,
+		HasProjectFilter: false,
+		HasContextFilter: false,
+		HasDueBefore:     false,
+		HasDue:           false,
+		HasDueAfter:      false,
+	}
+
+	dateParser := &DateParser{}
+
+	var subjectMatches []string
+
+	cr, _ := regexp.Compile(`\@[\p{L}\d_-]+`)
+	filter.Contexts = p.matchWords(input, cr)
+
+	pr, _ := regexp.Compile(`\+[\p{L}\d_-]+`)
+	filter.Projects = p.matchWords(input, pr)
+
+	for _, word := range strings.Split(input, " ") {
+		match := false
+		r1, _ := regexp.Compile(`archived:.*$`)
```

Comment:
I am assuming because we supply the string, and it doesn't come from outside, we feel safe there should be no errors.?

Reply:
`input_parser.go` deals with input directly from the user, so there could be errors!

However, for this specific regex, I'm failing to see what might error out.  The `Parse` function already returns an `error` so we could use it if need be, but I'd really need an example of a failing regex before we catch the error and return it.